### DISCUSSION TITLE
Redirects in a plugin generate hydration node mismatch

### DISCUSSION THREAD

**Discussion by david8z** [nuxt/nuxt#17874] [Questions] [answered]
How should redirects be implemented within a plugin?

I am using `$router.push({ path: '/login' })` which generates a Hydration node mismatch exception

**Comment by danielroe** [comment]
You should probably use route middleware for that.

**Reply by david8z** [reply]
I am kind of using a combination of both.

I am using firebase auth that has a listener called `onAuthStateChanged` so I declare that listener on a plugin.

Lets say the auth state changes to null, it will redirect it to '/login/'.

From what I see it doesn't make much sense to declare this on the middleware as it will be triggered on every route call, and I actually only want to declare it once.

Any tips or ideas on how this could be solved?

**Reply by david8z** [reply]
I just came across the hooks didn't knew about there existence does declaring my listener on them will make any more sense?

**Reply by david8z** [reply]
I ended up mutating the state and making the redirections in an function that was subscribed to the state changes. Nevertheless thanks for the help Daniel.

**Reply by david8z** [reply]
PD: how should I close this?

**Reply by danielroe** [reply]
You can add your solution as another reply and mark it as the answer.

**Comment by david8z** [comment, marked as answer]
> I ended up mutating the state and making the redirections in an function that was subscribed to the state changes.
